here-naksha-lib-model/src/commonMain/kotlin/naksha/model/TagNormalizer.kt

```diff
@@ -0,0 +1,129 @@
+ * | ref_     | NFKC       | false     | false    | false |
+ * | ~        | NFD        | false     | true     | true  |
+ * | #        | NFD        | false     | true     | true  |
+ * | sourceID | NFKC       | false     | false    | false |
+ * | < ELSE > | NFD        | true      | true     | true  |
+ * +----------+------------+-----------+----------+-------+
+ *
+ * By default, (if no special prefix is found) tag is normalized with NFD, lowercased, cleaned of non-ASCII and splittable.
+ */
+class TagNormalizer private constructor() {
+    private data class TagProcessingPolicy(
+        val normalizerForm: NormalizerForm,
+        val removeNonAscii: Boolean,
+        val lowercase: Boolean,
+        val split: Boolean
+    )
+
+    companion object TagNormalizer_C {
+        private val DEFAULT_POLICY =
+            TagProcessingPolicy(NFD, removeNonAscii = true, lowercase = true, split = true)
+        private val PREFIX_TO_POLICY = mapOf(
+            "@" to TagProcessingPolicy( NFKC, removeNonAscii = false, lowercase = false, split = true),
+            "ref_" to TagProcessingPolicy( NFKC, removeNonAscii = false, lowercase = false, split = false),
+            "sourceID" to TagProcessingPolicy( NFKC, removeNonAscii = false, lowercase = false, split = false),
+            "~" to TagProcessingPolicy(NFD, removeNonAscii = true, lowercase = false, split = true),
+            "#" to TagProcessingPolicy(NFD, removeNonAscii = true, lowercase = false, split = true)
+        )
+
+        private val AS_IS: CharArray = CharArray(128 - 32) { (it + 32).toChar() }
+        private val TO_LOWER: CharArray = CharArray(128 - 32) { (it + 32).toChar().lowercaseChar() }
+
+        /**
+         * Main method for raw tag normalization. See[TagNormalizer] doc for more
+         */
+        fun normalizeTag(tag: String): String {
+            val policy = policyFor(tag)
+            val normalized = Platform.normalize(tag, policy.normalizerForm)
+            return if (policy.lowercase) {
+                if (policy.removeNonAscii) {
+                    removeNonAscii(normalized, TO_LOWER)
+                } else {
+                    normalized.lowercase()
+                }
+            } else if (policy.removeNonAscii){
+                removeNonAscii(normalized, AS_IS)
+            } else {
+                normalized
+            }
+        }
+
+        private fun removeNonAscii(input: String, outputCharacterSet: CharArray): String {
+            val sb = StringBuilder()
+            for (element in input) {
+                val c = (element.code - 32).toChar()
+                if (c.code < outputCharacterSet.size) {
+                    sb.append(outputCharacterSet[c.code])
+                }
+            }
+            return sb.toString()
+        }
+
+
+        /**
+         * Main method for normalized tag splitting. See[TagNormalizer] doc for more
+         */
+        fun splitNormalizedTag(normalizedTag: String): Pair<String, Any?> {
```

Comment:
Pair is not exportable to JS, do we need to keep the method public? If not, just make it private or internal, this should solve the issue I would say. Otherwise, you can use `kotlin.collections.Map.Entry` or its implementation?

Reply:
`internal` should do the trick - changed